ando/tools/generator/bidsconverter.py

```diff
@@ -1,76 +1,318 @@
+import json
+import shutil
 from abc import ABC, abstractmethod
 from collections import defaultdict
 from pathlib import Path
 
+import pandas as pd
+from tqdm import tqdm
+
+from ando.AnDOChecker import is_valid
+
+REQ_FILES = dict(
+    description="dataset_description.json",
+    participant="participants.tsv",
+    session="{subject_label}\\{subject_label}_sessions.tsv",
+    channels="{subject_label}\\{session_label}\\ephys\\{subject_label}_{session_label}_channels.tsv",
+    contacts="{subject_label}\\{session_label}\\ephys\\{subject_label}_{session_label}_contacts.tsv",
+    ephys="{subject_label}\\{session_label}\\ephys\\{subject_label}_{session_label}_ephys.json",
+    probes="{subject_label}\\{session_label}\\ephys\\{subject_label}_{session_label}_probes.tsv",
+    file="{subject_label}\\{session_label}\\ephys\\{subject_label}_{session_label}_ephys{file_type}",
+)
+
 
 class BidsConverter(ABC):
+    file_type = ""
 
     def __init__(self, dataset_path, **kwargs):
         self.dataset_path = Path(dataset_path)
+        self.output_path = self.dataset_path.parent / "BIDSExt" / self.dataset_path.name
         self._kwargs = kwargs
-        self._participants_dict = dict(name=Path('participants.tsv'),
-                                       data=None)
-        self._dataset_desc_json = dict(name=Path('dataset_description.json'),
-                                       data=None)
+        self._participants_dict = dict()
+        self._dataset_desc_json = dict()
         self._sessions_dict = defaultdict(dict)
         self._channels_dict = defaultdict(dict)
         self._contacts_dict = defaultdict(dict)
         self._ephys_dict = defaultdict(dict)
         self._probes_dict = defaultdict(dict)
-        self._nwbfile_name_dict = defaultdict(dict)
-        self.datafiles_list = []
+        self._file_name_dict = defaultdict(dict)
+        self.datafiles_list = list(self.dataset_path.glob(f"**/*{self.file_type}"))
+        assert len(self.datafiles_list) > 0, f"no files of type {self.file_type} found"
+        self._datafiles_io = self._datafiles_open()
+        self._datafiles = [file.read() for file in self._datafiles_io]
+        self._labels_dict = defaultdict(list)
+        self.extract_metadata()
+
+    @abstractmethod
+    def _datafiles_open(self):
+        pass
+
+    def __del__(self):
+        for file in self._datafiles_io:
+            file.close()
+
+    def _tqdm(self, tqdm_obj, desc):
+        t = tqdm(tqdm_obj)
+        t.set_description(desc)
+        return t
```

Comment:
why did you make this a method? Why not just call e.g. 

```python
tqdm(self._datafiles, desc="participant")
```

Reply:
oops! yeah, no reason to make a method here. Didn't know this could be done directly too. Thanks for catching that